cmd/migration-managerd/internal/api/workers.go

```diff
@@ -45,251 +45,355 @@ func (d *Daemon) runPeriodicTask(f func() bool, interval time.Duration) {
+func (d *Daemon) trySyncAllSources(ctx context.Context, skipNonResponsiveSources bool) error {
+	log := slog.With(slog.String("method", "syncAllSources"))
+	var sources migration.Sources
 	err := transaction.Do(ctx, func(ctx context.Context) error {
 		// Get the list of configured sources.
-		sources, err := d.source.GetAll(ctx)
+		var err error
+		sources, err = d.source.GetAll(ctx)
 		if err != nil {
 			return fmt.Errorf("Failed to get all sources: %w", err)
 		}
 
-		// Check each source for any net networks and any new, changed, or deleted instances.
-		for _, src := range sources {
-			log := log.With(slog.String("source", src.Name))
+		return nil
+	})
+	if err != nil {
+		return err
+	}
 
-			s, err := source.NewInternalVMwareSourceFrom(api.Source{
-				Name:       src.Name,
-				DatabaseID: src.ID,
-				Insecure:   src.Insecure,
-				SourceType: src.SourceType,
-				Properties: src.Properties,
-			})
-			if err != nil {
-				log.Warn("Failed to create VMwareSource from source", logger.Err(err))
+	networksBySrc := map[string]map[string]api.Network{}
+	instancesBySrc := map[string]map[uuid.UUID]migration.Instance{}
+	for _, src := range sources {
+		srcNetworks, srcInstances, err := fetchVMWareSourceData(ctx, src)
+		if err != nil {
+			if skipNonResponsiveSources {
+				log.Error("Failed to fetch records from source", logger.Err(err))
 				continue
 			}
 
-			err = s.Connect(d.ShutdownCtx)
+			return err
+		}
+
+		networksBySrc[src.Name] = srcNetworks
+		instancesBySrc[src.Name] = srcInstances
+	}
+
+	var dbNetworks migration.Networks
+	var dbInstances migration.Instances
+	return transaction.Do(ctx, func(ctx context.Context) error {
+		dbNetworks, err = d.network.GetAll(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed to get internal network records: %w", err)
+		}
+
+		dbInstances, err = d.instance.GetAll(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed to get internal instance records: %w", err)
+		}
+
+		// Build maps to make comparison easier.
+		existingNetworks := make(map[string]migration.Network, len(dbNetworks))
+		for _, net := range dbNetworks {
+			existingNetworks[net.Name] = net
+		}
+
+		existingInstances := make(map[uuid.UUID]migration.Instance, len(dbInstances))
+		for _, inst := range dbInstances {
+			existingInstances[inst.UUID] = inst
+		}
+
+		for srcName, srcInstances := range instancesBySrc {
+			srcNetworks := networksBySrc[srcName]
+			err = syncNetworksFromSource(ctx, srcName, d.network, existingNetworks, srcNetworks)
 			if err != nil {
-				log.Warn("Failed to connect to source", logger.Err(err))
-				continue
+				return err
 			}
 
-			networks, err := s.GetAllNetworks(d.ShutdownCtx)
+			err = syncInstancesFromSource(ctx, srcName, d.instance, existingInstances, srcInstances)
 			if err != nil {
-				log.Warn("Failed to get networks", logger.Err(err))
-				continue
+				return err
 			}
+		}
```

Comment:
This looks pretty similar to `syncSourceData`, no? Can these two code sections be combined?

Reply:
Added a shared helper. Sadly to maintain the same behaviour, the signature of the shared helper is pretty long (two maps of maps).